File: day07/src/main/cpp/solution.cpp
#include <fstream>
#include <iostream>
#include <vector>

using namespace std;

vector<int> readInput() {
  ifstream file("../resources/input");
  vector<int> positions;
  for (int i; file >> i;) {
    positions.push_back(i);
    if (file.peek() == ',') {
      file.ignore();
    }
  }
  return positions;
}

long long moveCrabsCheapFuel(vector<int> const &crabs) {
  long long best = __LONG_LONG_MAX__;
  int min = INT32_MAX;
  int max = INT32_MIN;
  for (auto const &crab : crabs) {
    if (crab < min) {
      min = crab;
    }
    if (crab > max) {
      max = crab;
    }
  }
  for (int i = min; i <= max; i++) {
    long long count = 0;
    for (auto const &crab : crabs) {
      count += i < crab ? crab - i : i - crab;
    }
    if (count < best) {
      best = count;
    }
  }
  return best;
}

long long moveCrabsExpensiveFuel(vector<int> const &crabs) {
  long long best = __LONG_LONG_MAX__;
  int min = INT32_MAX;
  int max = INT32_MIN;
  for (auto const &crab : crabs) {
    if (crab < min) {
      min = crab;
    }
    if (crab > max) {
      max = crab;
    }
  }
  for (int i = min; i <= max; i++) {
    long long count = 0;
    long long diff;
    for (auto const &crab : crabs) {
      diff = i < crab ? crab - i : i - crab;
      count += (diff * (diff + 1)) / 2;
    }
    if (count < best) {
      best = count;
    }
  }
  return best;
}

int main() {
  vector<int> crabs = readInput();
  cout << "part 1" << endl
       << moveCrabsCheapFuel(crabs) << endl
       << "part 2" << endl
       << moveCrabsExpensiveFuel(crabs) << endl;
  return 0;
}

File: day07/src/main/python/solution.py
def read_input():
    f = open('../resources/input', 'r')
    return list(map(lambda x: int(x), f.readline().split(",")))


def find_fuel_simple(input):
    best = 9999999999999999999999999999999999999999
    for i in range(min(input), max(input)):
        count = sum(map(lambda p: p - i if p >= i else i - p, input))
        if count < best:
            best = count
    return best


def find_fuel_proper(input):
    best = 9999999999999999999999999999999999999999
    for i in range(min(input), max(input)):
        count = sum(map(lambda d: (d*(d+1))//2, map(lambda p: p - i if p >= i else i - p, input)))
        if count < best:
            best = count
    return best


if __name__ == "__main__":
    crabs = read_input()
    print("part 1")
    print(find_fuel_simple(crabs))
    print("part 2")
    print(find_fuel_proper(crabs))
